craft_store/auth.py

```diff
@@ -0,0 +1,87 @@
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Craft Store Authentication Store."""
+
+import base64
+import logging
+
+import keyring
+import keyring.errors
+
+from . import errors
+
+logger = logging.getLogger(__name__)
+
+
+class Auth:
+    """Auth wraps around the keyring to store credentials.
+
+    The application_name and host are used as key/values in the keyring to set,
+    get and delete credentials.
+
+    Credentials are base64 encoded into the keyring and decoded on retrieval.
+
+    :ivar application_name: name of the application using this library.
+    :ivar host: specific host for the store used.
+    """
+
+    def __init__(self, application_name: str, host: str) -> None:
+        """Initialize Auth.
+
+        :param application_name: name of the application using this library.
+        :param host: specific host for the store used.
+        """
+        self.application_name = application_name
+        self.host = host
+
+    def set_auth(self, auth: str) -> None:
+        """Store credentials in the keyring.
+
+        :param auth: token to store.
+        """
+        logger.debug(
+            "Storing credentials for %r on %r in keyring.",
+            self.application_name,
+            self.host,
+        )
+        encoded_auth = base64.b64encode(auth.encode())
+        keyring.set_password(self.application_name, self.host, encoded_auth.decode())
+
+    def get_auth(self) -> str:
+        """Retrieve credentials from the keyring."""
+        logger.debug(
+            "Retrieving credentials for %r on %r from keyring.",
+            self.application_name,
+            self.host,
+        )
+        encoded_auth_string = keyring.get_password(self.application_name, self.host)
+        if encoded_auth_string is None:
+            raise errors.NotLoggedIn()
+        auth = base64.b64decode(encoded_auth_string).decode()
+        return auth
+
+    def del_auth(self) -> None:
+        """Delete credentials from the keyring."""
+        logger.debug(
+            "Deleting credentials for %r on %r from keyring.",
+            self.application_name,
+            self.host,
+        )
+        try:
+            keyring.delete_password(self.application_name, self.host)
+        except keyring.errors.PasswordDeleteError as delete_error:
+            raise errors.NotLoggedIn() from delete_error
```

Comment:
Can't this be a noop? Maybe with a debug message in this case. But not failing.

Reply:
Can password delete errors happen due to other reasons that are not authentication errors (such as the keyring being locked by a different operation or something like that)? If so, we shouldn't assume it's always a `NotLoggedIn` error.

Reply:
this was the reason,

https://github.com/canonical/charmcraft/issues/191

I suppose the application could catch it and warn at that point

Reply:
for locked and no keyring there are keyring.errors.KeyringLocked and keyring.errors.NoKeyringError I am letting those float around until we have a clearer scenario (all methods should be decorated with this.)